core/types.ml

```diff
@@ -649,282 +539,305 @@ let primary_kind_of_type_arg : type_arg -> PrimaryKind.t =
-      (* | `Application ({Abstype.id="List"}, [`Type t]) -> is_unl_type (rec_vars, quant_vars) t  *)
-      | `Application _ -> true (* TODO: change this if we add linear abstract types *)
-      | `RecursiveApplication { r_linear ; _ } ->
-          (* An application is linear if the type it refers to is
-           * also linear. We calculate this information in two stages.
-           * The first pass (if r_linear () returns None) calculates linearity
-           * *up to recursive applications*, under the assumption that every type in the
-           * block is unrestricted. With this in hand, we can calculate
-           * linearity information, meaning that (r_linear ()) will return (Some lin). *)
-          OptionUtils.opt_app (not) true (r_linear ())
-      | `MetaTypeVar point -> is_unl_point is_unl_type (rec_vars, quant_vars) point
-      | `ForAll (qs, t) -> is_unl_type (rec_vars, add_quantified_vars !qs quant_vars) t
-      | `Dual s -> is_unl_type (rec_vars, quant_vars) s
-      | `End -> false
-      | #session_type -> false
-and is_unl_field vars =
-  function
-  | `Absent -> true
-  | `Present t ->
-    is_unl_type vars t
-  | `Var point -> is_unl_point is_unl_field vars point
-and is_unl_row vars (fields, row_var, _) =
-  let unl_row_var = is_unl_point is_unl_row vars row_var in
-  let unl_fields =
-    FieldEnv.fold
-      (fun _ f b -> b && is_unl_field vars f)
-      fields
-      true in
-  unl_row_var && unl_fields
+  val can_type_be : typ -> bool
+  val can_row_be : row -> bool
 
-let point_can_be_unl =
-  fun f ((rec_vars, quant_vars) as vars) point ->
-    begin
-        match Unionfind.find point with
-        | `Closed -> true
-        | `Var (v, (lin, _), `Rigid) -> IntSet.mem v quant_vars || Linearity.is_nonlinear lin
-        | `Var (_, _, `Flexible)     -> true
-        | `Body t -> f vars t
-        | `Recursive (var, t) ->
-          check_rec var rec_vars true (fun rec_vars' -> f (rec_vars', quant_vars) t)
-      end
+  val make_type : typ -> unit
+  val make_row : row -> unit
+end
 
-let rec type_can_be_unl : var_set * var_set -> typ -> bool =
-  fun ((rec_vars, quant_vars) as vars) ->
-    let tcu t = type_can_be_unl vars t in
-    function
+type predicate_context = StringSet.t * var_set * var_set
+
+class virtual type_predicate = object(self)
+  method is_var : (int * subkind * freedom) -> bool = fun _ -> true
+
+  method is_point : 'a 'c . (predicate_context -> 'a -> bool) -> predicate_context -> ([< 'a meta_max_basis] as 'c) point -> bool
+    = fun f ((rec_appl, rec_vars, quant_vars) as vars) point ->
+    match Unionfind.find point with
+    | `Closed -> true
+    | `Var ((id, _, _) as var) -> IntSet.mem id quant_vars || self#is_var var
+    | `Body t -> f vars t
+    | `Recursive (var, t) -> check_rec var rec_vars true (fun rec_vars' -> f (rec_appl, rec_vars', quant_vars) t)
+
+  method is_type ((rec_appl, rec_vars, quant_vars) as vars) : typ -> bool = function
```

Comment:
Checking a `typ` to see whether it is a type seems counterintuitive. Could you clarify when this would return false (and add as a comment?)

Reply:
Would it be better to call these methods `x_satisfies` instead of `is_x` (so `type_satisfies`, etc...)? Might be a little clearer what we're referring to.

Reply:
Oh definitely -- that makes it a lot clearer.